usr/src/uts/common/fs/zfs/vdev_indirect.c

```diff
@@ -1099,6 +1146,65 @@ vdev_indirect_io_start(zio_t *zio)
 	zio_execute(zio);
 }
 
+/* ARGSUSED */
+static void
+vdev_indirect_io_done(zio_t *zio)
+{
+	indirect_vsd_t *iv = zio->io_vsd;
+
+	if (!iv->iv_split_block) {
+		/*
+		 * This was not a split block, so we passed the BP down,
+		 * and the checksum was handled by the (one) child zio.
+		 */
+		return;
+	}
+	if (iv->iv_mirror_max_children == 0) {
+		/*
+		 * We didn't point to any mirrors, in which case there's
+		 * nothing else to try.  The checksum will be verified
+		 * by this zio.
+		 */
+		return;
+	}
+
+	zio_bad_cksum_t zbc;
+	int ret = zio_checksum_error(zio, &zbc);
+	if (ret == 0)
+		goto out;
+
+	if (iv->iv_mirror_child_hint + 1 < iv->iv_mirror_max_children) {
+		/*
+		 * The checksum doesn't match, but there are more mirrored
+		 * copies of this data.  Try a different version.
+		 */
+		iv->iv_mirror_child_hint++;
+		zio_vdev_io_redone(zio);
+		vdev_indirect_remap(zio->io_vd, zio->io_offset, zio->io_size,
+		    vdev_indirect_io_start_cb, zio);
+                return;
+	} else {
+		zio->io_error = ret;
+	}
+
+out:
+	/*
+	 * Note, indirect vdevs don't handle a few important cases for split
+	 * blocks when one side of a mirror is silently corrupt:
+	 *
+	 * When we read bad data and then retry the other side of the mirror,
+	 * we should issue a repair write to fix the bad copy.
+	 *
+	 * When scrubbing, we should read all copies of the data (and
+	 * repair any bad copies).
+	 *
+	 * We should generate ereports for any data we read with the wrong
+	 * checksums.
+	 */
+
+	zio_checksum_verified(zio);
```

Comment:
In the case where we are out of mirrors to check, but have not found a good copy of the data, do we want to mark the checksum as having been verified by unsetting the ZIO_STAGE_CHECKSUM_VERIFY bit in zio->io_pipeline here? Does this save us re-checking the checksum again later, and finding that it still does not match?

Or should we skip this if we have set zio->io_error already?

Reply:
Since we've already verified that the checksum is wrong, and set `io_error` appropriately, there isn't any point to doing it again, which is what calling zio_checksum_verified() accomplishes.  This is similar to what happens near the end of `vdev_raidz_io_done()`.